[14.0][BPRT] edi_sale_ubl_oca

thienvh332 opened the pull request.

simahawk committed:
Add edi_sale_oca

simahawk committed:
edi_sale: take over edi_sale_order_import

simahawk committed:
edi_sale_order_import: deprecate

simahawk committed:
Add edi_sale_ubl_oca

simahawk committed:
edi_sale: filter/group by origin

simahawk committed:
edi_sale: handle ext ref for sol

simahawk committed:
edi_sale_ubl: improve order response data

simahawk committed:
edi_sale_ubl_oca: make exc type demo FIXME

simahawk committed:
edi_sale_ubl_oca: depend on party data

simahawk committed:
edi_sale_ubl_oca: move data to demo

simahawk committed:
edi_sale_ubl_oca: imp OrderResponse data

simahawk committed:
edi_sale_ubl_oca: fix tests

simahawk committed:
edi_sale_ubl: ease data override via info

simahawk committed:
edi_sale: show edi_id

simahawk committed:
edi_sale: store metadata

simahawk committed:
edi_sale: clean file and drop *type_auto dependency

simahawk committed:
edi_sale_ubl: fix demo data

simahawk committed:
edi_sale_ubl: introduce states

simahawk committed:
edi_sale: add process hook for order create

simahawk committed:
edi_sale: process propagate edi origin

simahawk committed:
edi_sale_ubl: improve tests common

simahawk committed:
edi_sale: add edi_exchange_ready

simahawk committed:
edi_sale: add edi_exchange_ready

simahawk committed:
edi_sale_ubl: fix state mapping

simahawk committed:
edi_sale_ubl: fix order state handling

simahawk committed:
edi_sale_ubl: do not use Form for tests

The Form class is just too dumb.
If a field is not there it's broken
and if you pass a view for a main record
you cannot pass any for the x2m fields
which makes it impossible to use for general purpose.

simahawk committed:
edi_sale_ubl: fix state update

simahawk committed:
edi_sale: fix edi disable flag

simahawk committed:
Add edi_sale_endpoint

simahawk committed:
edi_sale_ubl: improve line state test

simahawk committed:
edi_sale_ubl: trigger state update on line create/write

simahawk committed:
edi_sale_ubl: fix line state compute

Additionally, avoid recomputing states for all lines
when we know which lines have changed.

simahawk committed:
edi_sale_ubl: fix order state compute

Several lines' state make the order not satisfied.

simahawk committed:
edi_sale_ubl: adapt to type rule

simahawk committed:
edi_sale_ubl: make state readonly

simahawk committed:
edi_sale: fix _edi_exchange_ready for qty=0

Lines w/ no qty must not be excluded a priori

simahawk committed:
edi_sale_ubl: fix line state compute

simahawk committed:
edi_sale_ubl: fix tests order setup

simahawk committed:
edi_sale_ubl: fix line state compute on order create

simahawk committed:
edi_sale_ubl: fix ordrsp tmpl

simahawk committed:
edi_sale_ubl: fix line state for qty=0

simahawk committed:
edi_sale_ubl: fix tests, handle increased qty

simahawk committed:
edi_sale: move OrderMixin from edi_sale_ubl

simahawk committed:
edi_sale: move OrderMixin from edi_sale_ubl

simahawk committed:
edi_sale: fix test_process

simahawk committed:
edi_sale: propagate origin on lines

Commit:
[MIG] edi_sale_oca: Migration to 16.0

Commit:
[IMP] edi_sale_oca: Support edi_id for edi_sale_edifact_oca

Commit:
[IMP] edi_sale_oca: Refactor code for passing all configurations via env_ctx

The configurations here are values for the wizard and whatever we need

Commit:
[UPD] Update edi_sale_oca.pot

OCA-git-bot committed:
[BOT] post-merge updates

Commit:
[IMP] edi_oca: add partner form page

Commit:
[UPD] Update edi_sale_oca.pot

OCA-git-bot committed:
[BOT] post-merge updates

Commit:
[FIX] edi_sale_oca: fix test

Commit:
[UPD] Update edi_sale_oca.pot

OCA-git-bot committed:
[BOT] post-merge updates

Commit:
[UPD] Update edi_sale_oca.pot

mymage committed:
Added translation using Weblate (Italian)

Commit:
[IMP] edi_sale_ubl_oca: add more decorator on create method

thienvh332 committed:
[BPRT] edi_sale_oca: Backport from v16.0

thienvh332 committed:
[IMP] edi_sale_oca: add partner form page

QuocDuong1306 commented on edi_sale_ubl_oca/tests/test_order_in.py:
```diff
@@ -60,6 +68,18 @@ def test_new_order(self):
+                ]
+            )
+            order.update({"origin_exchange_record_id": res[0]})
+            order.order_line.update({"origin_exchange_record_id": res[0]})
```
Did you take into account for [this commit](https://github.com/OCA/edi/pull/831/commits/ec3f3722177b45b9f32070715226a78926138b17)?
And also need the same commits for [edi_sale_order_import_ubl](https://github.com/OCA/edi/blob/01dfaaf2f2067af3e71ac27696d6dd828b8e2d74/edi_sale_order_import_ubl/__manifest__.py#L15) and [edi_sale_order_import_ubl_endpoint](https://github.com/OCA/edi/blob/01dfaaf2f2067af3e71ac27696d6dd828b8e2d74/edi_sale_order_import_ubl_endpoint/__manifest__.py#L15)?

QuocDuong1306 commented on edi_sale_ubl_oca/tests/test_order_in_full_flow.py:
```diff
@@ -5,23 +5,31 @@
+    def setUp(self):
+        super().setUp()
+        ComponentMixin.setUp(self)
+        self._setup_inbound_order(self.backend)
```
Move `cls._setup_inbound_order(cls.backend)` to `setUpClass`, and you can delete `def setUp`

QuocDuong1306 commented on edi_sale_ubl_oca/tests/test_order_response_out.py:
```diff
@@ -76,4 +82,4 @@ def test_xml(self):
         data = handler.parse_xml(file_content)
         # TODO: test all main data
-        self.assertEqual(data["cbc:OrderResponseCode"], "AP")
+        self.assertEqual(data["cbc:OrderResponseCode"], "CA")
```
Can put a comment something like `Order has not been created from exc record in (just linked to record out), so all lines are not in Accepted state`

You can see [here](https://github.com/OCA/edi-framework/blob/669c46b9d44291c59aa85bc8a95ca83f72e97c5f/edi_record_metadata_oca/models/edi_exchange_consumer_mixin.py#L23) --> no store metadata to change the state in lines

QuocDuong1306 commented on edi_sale_ubl_oca/tests/test_order_response_out.py:
```diff
@@ -36,6 +37,11 @@ def setUpClass(cls):
         cls.sale.order_line.origin_exchange_record_id = cls.record
         cls.sale._edi_update_state()
 
+    def setUp(self):
+        super().setUp()
+        ComponentMixin.setUp(self)
+        self.sale.action_confirm()
+
```
Maybe no need this one

QuocDuong1306 commented on edi_sale_ubl_oca/tests/test_order_in_full_flow.py:
```diff
@@ -48,6 +56,7 @@ def test_new_order(self):
         ack_exc_record = order.exchange_record_ids.filtered(
             lambda x: x.type_id == self.exc_type_out
         )
+        ack_exc_record.action_exchange_generate()
```
Could you explain a little bit, why we have set `exchange_file_auto_generate` already, but still need to call `action_exchange_generate` here (just asking)

thienvh332 replied:
Thanks for your review @QuocDuong1306 
After *_auto is created automatically, it will call the triggers. But if the triggers do not exist, nothing will happen.
You can see it [here](https://github.com/camptocamp/edi-framework/blob/4e14fb3d03abc7e500c31dd4165209437fb46783/edi_exchange_type_auto/models/edi_auto_exchange_consumer_mixin.py#L387-L389)

QuocDuong1306 replied:
Got it, need to define `on_edi_auto_generate` or `auto_handle_generate` events to use the `auto` type rule

QuocDuong1306 reviewed (commented):
Hi a @thienvh332 , I have put some reviews, could you double-check again?

alexis-via committed:
[FIX] account_invoice_import_simple_pdf: version extraction for fitz/PyMuPDF

Use with statement for NamedTemporaryFile()

alexis-via committed:
[FIX] account_invoice_import_simple_pdf: add seek(0) to avoid error "empty file"

Add try/except on pypdf text extraction

QuocDuong1306 commented on edi_sale_ubl_oca/demo/edi_exchange_type.xml:
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+<!-- TODO: noupdate OR better create via post init hook   -->
+    <record id="demo_edi_exc_type_order_response_out" model="edi.exchange.type">
+        <field name="backend_type_id" ref="edi_ubl_oca.edi_backend_type_ubl" />
+        <field name="name">Demo UBL Sale Order Response</field>
+        <field name="code">demo_UBL_SaleOrderResponse_out</field>
+        <field name="direction">output</field>
+        <field name="exchange_filename_pattern">{record_name}-{type.code}-{dt}</field>
+        <field name="exchange_file_ext">xml</field>
```
Did we miss `backend_id`?

thienvh332 replied:
I don't know why simone didn't declare backend_id here but added it in [unit test](https://github.com/thienvh332/edi/blob/2b961b6a40f11f33214821200ff5d03bc9ddebf7/edi_sale_ubl_oca/tests/common.py#L65-L66) in the same commit.

QuocDuong1306 replied:
Please ignore
Actually, if backend is not defined, `create_record` will be based on `backend_type_id` ([here](https://github.com/OCA/edi/blob/e9fe1a70f7db810dfd5da879b8f25a15f490951a/edi_oca/models/edi_backend.py#L187)), so ok for me

QuocDuong1306 replied:
But take note that: `Auto` won't be used if not defined `backend_id` ([here](https://github.com/camptocamp/edi-framework/blob/4e14fb3d03abc7e500c31dd4165209437fb46783/edi_exchange_type_auto/models/edi_auto_exchange_consumer_mixin.py#L113))

QuocDuong1306 commented on edi_sale_ubl_oca/demo/edi_exchange_type.xml:
```diff
@@ -0,0 +1,75 @@
+        </field>
+    </record>
+
+    <record id="demo_edi_exc_type_order_in" model="edi.exchange.type">
+        <field name="backend_type_id" ref="edi_ubl_oca.edi_backend_type_ubl" />
+        <field name="ack_type_id" ref="demo_edi_exc_type_order_response_out" />
+        <field name="name">Demo UBL Sale Order</field>
+        <field name="code">demo_UBL_SaleOrder_in</field>
```
Missed `backend_id`?

thienvh332 replied:
Same here =))

QuocDuong1306 reviewed (approved):
Pre-approving

Do we need to change a little bit for commit's description?

![image](https://github.com/user-attachments/assets/8530e241-1efa-4bc2-b940-6889bc715c52)

QuocDuong1306 reviewed (approved):
`[IMP] edi_sale_ubl_oca: depend on edi_exchange_type_auto`

thienvh332 committed:
[FIX] pdf_helper: fix issue pypdf 5.0 dropped support for python<=3.7

alexis-via committed:
[FIX] account_invoice_download: don't set default_name in context

If default_name is set in context just to have a default label on the
account.invoice.import.download, it will also be used if do a manual run
right after, so you will get draft invoices with a 'name' already set !

OCA-git-bot committed:
Merge PR #1041 into 14.0

Signed-off-by simahawk

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
Merge PR #1018 into 14.0

Signed-off-by alexis-via

OCA-git-bot committed:
[BOT] post-merge updates

alexis-via committed:
[FIX] pypdf version for tests

OCA-git-bot committed:
Merge PR #1043 into 14.0

Signed-off-by alexis-via

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
Merge PR #1023 into 14.0

Signed-off-by simahawk

Commit:
[UPD] Update edi_oca.pot

Commit:
[UPD] Update edi_sale_oca.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: edi-14.0/edi-14.0-edi_oca
Translate-URL: https://translation.odoo-community.org/projects/edi-14-0/edi-14-0-edi_oca/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: edi-14.0/edi-14.0-edi_sale_oca
Translate-URL: https://translation.odoo-community.org/projects/edi-14-0/edi-14-0-edi_sale_oca/

OCA-git-bot committed:
Merge PR #1066 into 14.0

Signed-off-by simahawk

Commit:
[UPD] Update edi_sale_order_import.pot

simahawk committed:
test-requirements.txt: depend on edi.config

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: edi-14.0/edi-14.0-edi_sale_order_import
Translate-URL: https://translation.odoo-community.org/projects/edi-14-0/edi-14-0-edi_sale_order_import/

OCA-git-bot committed:
Merge PR #1022 into 14.0

Signed-off-by simahawk

Commit:
[UPD] Update edi_sale_endpoint.pot

OCA-git-bot committed:
[BOT] post-merge updates

simahawk committed:
edi_sale_ubl_oca: use edi.config

simahawk committed:
fixup! edi_sale_ubl_oca: use edi.config

simahawk committed:
fixup! fixup! edi_sale_ubl_oca: use edi.config

simahawk committed:
fixup! fixup! fixup! edi_sale_ubl_oca: use edi.config
